rr_evgp/src/costmap_plugins/track_closing_layer.cpp

```diff
@@ -0,0 +1,248 @@
+    }
+
+    using WallEndpointContext = std::tuple<cv::Point, cv::Point, double, double>;
+    bool IsConnectable(const WallEndpointContext& current, const WallEndpointContext& prev);
+};
+
+PLUGINLIB_EXPORT_CLASS(rr::TrackClosingLayer, costmap_2d::Layer)
+
+// ******************* Implementation ********************** //
+
+void TrackClosingLayer::onInitialize() {
+    dsrv_ = std::make_unique<dynamic_reconfigure::Server<costmap_2d::GenericPluginConfig>>();
+    dsrv_->setCallback(boost::bind(&TrackClosingLayer::reconfigureCB, this, _1, _2));
+
+    ros::NodeHandle private_nh("~" + getName());
+    assertions::getParam(private_nh, "branch_pruning_size", branch_pruning_size_, { assertions::greater<int>(0) });
+    assertions::getParam(private_nh, "dilate_size", dilate_size_, { assertions::greater<int>(0) });
+    assertions::getParam(private_nh, "min_enclosing_radius", min_enclosing_radius_, { assertions::greater<double>(0) });
+    assertions::getParam(private_nh, "reg_bounding_box_size", reg_bounding_box_size_, { assertions::greater<int>(0) });
+    assertions::getParam(private_nh, "extrapolate_distance", extrapolate_distance_, { assertions::greater<double>(0) });
+    assertions::getParam(private_nh, "max_angle_between_lines", max_angle_between_lines_,
+                         { assertions::greater<double>(0) });
+    assertions::getParam(private_nh, "max_dist_between_lines", max_dist_between_lines_,
+                         { assertions::greater<double>(0) });
+}
+
+void TrackClosingLayer::updateBounds(double robot_x, double robot_y, double robot_yaw, double* min_x, double* min_y,
+                                     double* max_x, double* max_y) {
+    if (!enabled_) {
+        return;
+    }
+
+    const costmap_2d::Costmap2D* costmap = layered_costmap_->getCostmap();
+    *min_x = std::min(*min_x, costmap->getOriginX());
+    *min_y = std::min(*min_y, costmap->getOriginY());
+    *max_x = std::max(*max_x, costmap->getOriginX() + costmap->getSizeInMetersX());
+    *max_y = std::max(*max_y, costmap->getOriginY() + costmap->getSizeInMetersY());
+}
+
+template <int N>
+double norm(const cv::Vec<double, N>& x) {
+    return std::sqrt(x.dot(x));
+}
+
+cv::Point2d intersect_point_line(const cv::Point2d& p, const cv::Point2d& l1, const cv::Point2d& l2, bool bound) {
+    cv::Vec2d n = l2 - l1;
+    n /= norm(n);
+    double t = (p - l1).dot(n);
+    cv::Vec2d x = cv::Vec2d(l1) + n * t;
+
+    if (bound && (norm(x - cv::Vec2d(l1)) + norm(x - cv::Vec2d(l2)) > norm(l2 - l1))) {
+        x = (norm(p - l1) < norm(p - l2)) ? l1 : l2;
+    }
+
+    return x;
+}
+
+bool TrackClosingLayer::IsConnectable(const WallEndpointContext& current, const WallEndpointContext& prev) {
+    const auto& [endpntA, pntA, vxA, vyA] = current;
+    const auto& [endpntB, pntB, vxB, vyB] = prev;
+
+    double angle_between_lines = std::acos(vxA * vxB + vyA * vyB) * 180.0 / M_PI;
+    angle_between_lines = std::min(angle_between_lines, 180 - angle_between_lines);
```

Comment:
I kinda doubt that `min(angle, 180 - angle)` is what we want here. That implies that angle differences of 0 and 180 are equally good. We're looking for angles near 180 only, right?